### PR TITLE
Add DDL support for indices disabling

### PR DESCRIPTION
**Pull request by gczsjdy**
## What changes were proposed in this pull request?

To enable disabling indices using DDL.

Note that only 1 index can be disabled at a time. For example, to disable `indexA`, `indexB`:
Two statements: `Disable oindex indexA on table` & `Disable oindex indexB on table` are needed.
Due to supporting of multiple indices disabling at a time will cause other queries' parse error.

## How was this patch tested?

Test case in `IndexSelectionSuite`.

Out of date: <del>Test is currently not added, due to this actually depend on PR#596.</del>

### CODE REVIEW

**Comment by carsonwang** [comment]
can you add a test using the DDL?

**Comment by carsonwang** [comment]
@gczsjdy , we also want to ENABLE the index again after DISABLE. LIST all disabled index is also necessary.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 1]
```diff
@@ -649,3 +649,54 @@ case class OapCheckIndexCommand(
 
   }
 }
+
+/**
+ * Disable specific indices
+ */
+case class OapDisableIndexCommand(
+    indexNames: String,
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: nit: indexName

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 2]
```diff
@@ -649,3 +649,54 @@ case class OapCheckIndexCommand(
...
+          // TODO get `fs` outside of foreach() to boost
+          val fs = parent.getFileSystem(sparkSession.sparkContext.hadoopConfiguration)
+          if (fs.exists(new Path(parent, OapFileFormat.OAP_META_FILE))) {
+            val metaBuilder = new DataSourceMetaBuilder()
+            val m = OapUtils.getMeta(sparkSession.sparkContext.hadoopConfiguration, parent)
+            assert(m.nonEmpty)
+            val oldMeta = m.get
+            val existsIndexNameList = oldMeta.indexMetas.map(_.name)
+            val existsData = oldMeta.fileMetas
+            val indexNameList = indexNames.split(",").map(_.trim)
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: Please update this if we only support disable one index

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 3]
```diff
@@ -649,3 +649,54 @@ case class OapCheckIndexCommand(
...
+            val notExistsIndexNameList = indexNameList.filterNot(existsIndexNameList.contains(_))
+            val notExistsIndicesString = notExistsIndexNameList.mkString(", ")
+            if (!notExistsIndexNameList.isEmpty) {
+              if (!allowNotExists) {
+                throw new AnalysisException(
+                  s"""Index $notExistsIndicesString does not exist on
+                     | ${identifier.getOrElse(parent)}""".stripMargin)
+              } else {
+                logWarning(s"drop non-exists index $notExistsIndicesString")
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: nit: drop -> Disable

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 4]
```diff
@@ -649,3 +649,54 @@ case class OapCheckIndexCommand(
...
+    partitionSpec: Option[TablePartitionSpec]) extends RunnableCommand {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val qe = sparkSession.sessionState.executePlan(UnresolvedRelation(table))
+    qe.assertAnalyzed()
+    val relation = qe.optimizedPlan
+
+    relation match {
+      case LogicalRelation(HadoopFsRelation(fileCatalog, _, _, _, format, _), _, identifier)
+        if format.isInstanceOf[OapFileFormat] || format.isInstanceOf[ParquetFileFormat] =>
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: nit: need two more spaces

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexSelectionSuite.scala [hunk 5]
```diff
@@ -20,12 +20,11 @@ package org.apache.spark.sql.execution.datasources.oap.index
 import java.io.File
 
 import scala.collection.mutable.ArrayBuffer
-
 import org.apache.hadoop.fs.Path
 import org.scalatest.BeforeAndAfterEach
-
```
Comment (carsonwang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexSelectionSuite.scala]: Why removing these empty lines?

File: src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 6]
```diff
@@ -156,7 +156,9 @@ statement
         partitionSpec?                                                 #oapDropIndex
     | DISABLE SINDEX (IF EXISTS)? IDENTIFIER ON tableIdentifier
         partitionSpec?                                                 #oapDisableIndex
+    | ENABLE SINDEX IDENTIFIER                                         #oapEnableIndex
     | SHOW SINDEX (FROM | IN) tableIdentifier                          #oapShowIndex
+    | SHOW DISABLED SINDICES                                           #oapShowDisabledIndices
```
Comment (yueawang) [inline src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Why not just add a column (enable) to show index?? refer to breakpoint in gdb.

|Num|Type          |    Disp|**Enb**|Address                       |What                                          |
|-----|------------|--------|--------|------------------------|----------------------------------|
|1     |breakpoint |  keep  | **y**   |0x000000000040077a| in main() at hello_world.cpp:6 |
|2     |breakpoint |  keep  |    n      | PENDING                   | main + 5                                   |

Another thing is will we need "SHOW ENABLED SINDICES"?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala (hunk 4 above)
Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: Sorry, why?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala (hunk 2 above)
Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: Done, : )

File: src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 6 above)
Comment (gczsjdy) [reply src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Sounds good, this seems more elegant than the current implementation to add a new grammar.
If a column is added, `show enabled/disabled sindices` is not needed I think. 
@carsonwang What do you think?

Comment (carsonwang) [reply src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: I like this idea.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala (hunk 4 above)
Comment (carsonwang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: It is part of the `case` statement, right?

Comment (carsonwang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: All right, just ignore this.

**Comment by carsonwang** [comment]
LGTM. Merging.

**Pull request merged**